File: frontend/src/types/serviceApplication.ts
/**
 * 서비스 이용신청 관련 타입 정의
 */

// 플랜 타입
export type PlanType = '1_MONTH' | '6_MONTHS' | '1_YEAR'

// 신청 상태
export type ApplicationStatus =
  | 'PENDING'           // 신청완료 (입금대기)
  | 'PAYMENT_CONFIRMED' // 입금확인
  | 'APPROVED'          // 승인완료
  | 'REJECTED'          // 거절
  | 'CANCELLED'         // 취소

// 플랜 정보
export interface EnterprisePlanInfo {
  plan_type: PlanType
  name_ko: string
  price: number
  price_display: string
  duration_days: number
  discount: string | null
}

// 신청 응답 (사용자용)
export interface ServiceApplicationResponse {
  id: string
  applicant_email: string
  plan_type: PlanType
  plan_name_ko: string
  plan_amount: number
  status: ApplicationStatus
  subscription_start_date: string | null
  subscription_end_date: string | null
  created_at: string
  updated_at: string
}

// 신청 목록 응답 (사용자용)
export interface ServiceApplicationListResponse {
  applications: ServiceApplicationResponse[]
  current_subscription: {
    status: 'ACTIVE' | 'NONE' | 'EXPIRED'
    tier: string
    end_date: string | null
  } | null
}

// 신청 응답 (관리자용)
export interface ServiceApplicationAdminResponse extends ServiceApplicationResponse {
  user_id: string
  has_business_registration: boolean
  business_registration_file_name: string | null
  admin_memo: string | null
  processed_by: string | null
  processed_at: string | null
}

// 신청 목록 응답 (관리자용)
export interface ServiceApplicationAdminListResponse {
  applications: ServiceApplicationAdminResponse[]
  total: number
  page: number
  total_pages: number
}

// 입금 정보
export interface PaymentBankInfo {
  company_name: string
  business_number: string
  bank_name: string
  account_number: string
  account_holder: string
}

// 신청 완료 응답 (입금 안내 포함)
export interface PaymentInfoResponse {
  application_id: string
  plan_type: PlanType
  plan_name_ko: string
  plan_amount: number
  plan_amount_display: string
  bank_info: PaymentBankInfo
  message: string
}

// 관리자용 타입 별칭
export type ServiceApplicationAdmin = ServiceApplicationAdminResponse

// 엔터프라이즈 플랜 정보 (프론트엔드용)
export const ENTERPRISE_PLANS: EnterprisePlanInfo[] = [
  {
    plan_type: '1_MONTH',
    name_ko: '1개월',
    price: 300000,
    price_display: '300,000원',
    duration_days: 30,
    discount: null
  },
  {
    plan_type: '6_MONTHS',
    name_ko: '6개월',
    price: 1500000,
    price_display: '1,500,000원',
    duration_days: 180,
    discount: '17% 할인'
  },
  {
    plan_type: '1_YEAR',
    name_ko: '1년',
    price: 3000000,
    price_display: '3,000,000원',
    duration_days: 365,
    discount: '17% 할인'
  }
]

// 상태별 표시 정보
export const APPLICATION_STATUS_INFO: Record<ApplicationStatus, {
  label: string
  color: string
  textColor: string
  bgColor: string
}> = {
  PENDING: {
    label: '입금대기',
    color: 'text-yellow-400',
    textColor: 'text-yellow-400',
    bgColor: 'bg-yellow-400/10'
  },
  PAYMENT_CONFIRMED: {
    label: '입금확인',
    color: 'text-blue-400',
    textColor: 'text-blue-400',
    bgColor: 'bg-blue-400/10'
  },
  APPROVED: {
    label: '승인완료',
    color: 'text-green-400',
    textColor: 'text-green-400',
    bgColor: 'bg-green-400/10'
  },
  REJECTED: {
    label: '거절',
    color: 'text-red-400',
    textColor: 'text-red-400',
    bgColor: 'bg-red-400/10'
  },
  CANCELLED: {
    label: '취소',
    color: 'text-gray-400',
    textColor: 'text-gray-400',
    bgColor: 'bg-gray-400/10'
  }
}

File: backend/app/schemas/service_application.py
"""
서비스 이용신청 스키마
"""
from pydantic import BaseModel, EmailStr, Field
from typing import Optional, List, Literal
from datetime import datetime, date
from uuid import UUID


# 플랜 타입
PlanTypeEnum = Literal["1_MONTH", "6_MONTHS", "1_YEAR"]

# 상태 타입
ApplicationStatusEnum = Literal[
    "PENDING",           # 신청완료 (입금대기)
    "PAYMENT_CONFIRMED", # 입금확인
    "APPROVED",          # 승인완료
    "REJECTED",          # 거절
    "CANCELLED"          # 취소
]


# ============================================================
# 플랜 정보
# ============================================================
class EnterprisePlanInfo(BaseModel):
    """엔터프라이즈 플랜 정보"""
    plan_type: PlanTypeEnum
    name_ko: str
    price: int
    price_display: str
    duration_days: int
    discount: Optional[str] = None


class EnterprisePlansResponse(BaseModel):
    """엔터프라이즈 플랜 목록 응답"""
    plans: List[EnterprisePlanInfo]


# ============================================================
# 서비스 이용신청 - 사용자용
# ============================================================
class ServiceApplicationCreate(BaseModel):
    """서비스 이용신청 생성 요청"""
    applicant_email: EmailStr = Field(..., description="신청자 이메일")
    plan_type: PlanTypeEnum = Field(..., description="신청 플랜")
    # 파일은 multipart/form-data로 별도 처리


class ServiceApplicationResponse(BaseModel):
    """서비스 이용신청 응답"""
    id: UUID
    applicant_email: str
    plan_type: str
    plan_name_ko: str
    plan_amount: int
    status: ApplicationStatusEnum
    subscription_start_date: Optional[date] = None
    subscription_end_date: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ServiceApplicationListResponse(BaseModel):
    """서비스 이용신청 목록 응답 (사용자용)"""
    applications: List[ServiceApplicationResponse]
    current_subscription: Optional[dict] = None  # 현재 이용권 상태


class ServiceApplicationCreateResult(BaseModel):
    """서비스 이용신청 생성 결과"""
    id: UUID
    status: ApplicationStatusEnum
    plan_type: str
    plan_amount: int
    created_at: datetime
    message: str


# ============================================================
# 서비스 이용신청 - 관리자용
# ============================================================
class ServiceApplicationAdminResponse(BaseModel):
    """서비스 이용신청 응답 (관리자용)"""
    id: UUID
    user_id: UUID
    applicant_email: str
    plan_type: str
    plan_name_ko: str
    plan_amount: int
    status: ApplicationStatusEnum
    has_business_registration: bool  # 사업자등록증 첨부 여부
    business_registration_file_name: Optional[str] = None
    admin_memo: Optional[str] = None
    processed_by: Optional[UUID] = None
    processed_at: Optional[datetime] = None
    subscription_start_date: Optional[date] = None
    subscription_end_date: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ServiceApplicationAdminListResponse(BaseModel):
    """서비스 이용신청 목록 응답 (관리자용)"""
    applications: List[ServiceApplicationAdminResponse]
    total: int
    page: int
    total_pages: int


class ServiceApplicationStatusUpdate(BaseModel):
    """서비스 이용신청 상태 변경 요청"""
    status: Literal["PAYMENT_CONFIRMED", "APPROVED", "REJECTED"]
    admin_memo: Optional[str] = None
    subscription_start_date: Optional[date] = None  # APPROVED 시 필수
    subscription_end_date: Optional[date] = None    # APPROVED 시 필수


class ServiceApplicationStatusUpdateResult(BaseModel):
    """서비스 이용신청 상태 변경 결과"""
    id: UUID
    status: ApplicationStatusEnum
    message: str


# ============================================================
# 입금 정보
# ============================================================
class PaymentBankInfo(BaseModel):
    """입금 계좌 정보"""
    company_name: str = "코넥트"
    business_number: str = "686-19-02309"
    bank_name: str = "카카오뱅크"
    account_number: str = "3333-31-9041159"
    account_holder: str = "코넥트 / 박재준"


class PaymentInfoResponse(BaseModel):
    """입금 안내 응답"""
    application_id: UUID
    plan_type: str
    plan_name_ko: str
    plan_amount: int
    plan_amount_display: str
    bank_info: PaymentBankInfo
    message: str = "신청이 완료되었습니다. 입금 안내 이메일을 확인해주세요."
